src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleArray.cs

```diff
@@ -246,8 +257,18 @@ public static partial class JsonSerializer
             }
             else if (state.Current.IsDictionary)
             {
-                Debug.Assert(state.Current.ReturnValue != null);
-                IDictionary<string, TProperty> dictionary = (IDictionary<string, TProperty>)state.Current.JsonPropertyInfo.GetValueAsObject(state.Current.ReturnValue);
+                object dictionaryObject = null;
+                if (state.Current.ReturnValue != null)
+                {
+                    dictionaryObject = state.Current.ReturnValue;
+                }
+                else if (state.Current.TempDictionaryValues != null)
+                {
+                    dictionaryObject = state.Current.TempDictionaryValues;
+                }
+                Debug.Assert(dictionaryObject != null);
```

Comment:
I was going to say this assert has no value because GetValueAsObject will throw, but I see it may simply return the value.

In general we do not include asserts that simply bring forward the point of failure by 1 millisecond. Eg., I would not include this assert, which I saw elsewhere:
```
            Debug.Assert(Get != null);
            return Get((TClass)obj);
```